spark_client/services.py

```diff
@@ -0,0 +1,845 @@
+                       account in all namespaces
+            labels: filter to be applied to retrieve service account which match certain labels.
+        """
+        cmd = "get serviceaccount"
+
+        if labels is not None and len(labels) > 0:
+            cmd += " ".join([f" -l {label}" for label in labels])
+
+        namespace = " -A" if namespace is None else f" -n {namespace}"
+
+        all_service_accounts_raw = self.exec(cmd + namespace)
+
+        if isinstance(all_service_accounts_raw, str):
+            raise ValueError("Malformed output")
+
+        return all_service_accounts_raw["items"]
+
+    def get_secret(self, secret_name: str, namespace: str) -> Dict[str, Any]:
+        """Return the data contained in the specified secret.
+
+        Args:
+            secret_name: name of the secret
+            namespace: namespace where the secret is contained
+        """
+
+        try:
+            secret = self.exec(
+                f"get secret {secret_name} --ignore-not-found", namespace=namespace
+            )
+        except Exception:
+            raise NoResourceFound(secret_name)
+
+        if secret is None or len(secret) == 0 or isinstance(secret, str):
+            raise NoResourceFound(secret_name)
+
+        result = dict()
+        for k, v in secret["data"].items():
+            # k1 = k.replace(".", "\\.")
+            # value = self.kube_interface.exec(f"get secret {secret_name}", output=f"jsonpath='{{.data.{k1}}}'")
+            result[k] = base64.b64decode(v).decode("utf-8")
+
+        secret["data"] = result
+        return secret
+
+    def set_label(
+        self, resource_type: str, resource_name: str, label: str, namespace: str
+    ):
+        """Set label to a specified resource (type and name).
+
+        Args:
+            resource_type: type of the resource to be labeled, e.g. service account, rolebindings, etc.
+            resource_name: name of the resource to be labeled
+            namespace: namespace where the resource is
+        """
+
+        self.exec(
+            f"label {resource_type} {resource_name} {label}",
+            namespace=namespace,
+        )
+
+    def create(
+        self, resource_type: str, resource_name: str, namespace: str, **extra_args
+    ):
+        """Create a K8s resource.
+
+        Args:
+            resource_type: type of the resource to be created, e.g. service account, rolebindings, etc.
+            resource_name: name of the resource to be created
+            namespace: namespace where the resource is
+            extra_args: extra parameters that should be provided when creating the resource. Note that each parameter
+                        will be prepended with the -- in the cmd, e.g. {"role": "view"} will translate as
+                        --role view in the command
+        """
+
+        formatted_extra_args = " ".join([f"--{k}={v}" for k, v in extra_args.items()])
```

Comment:
this is cool! is it applicable to all k8s resources in general?

Reply:
yes, but it is neither type-safe nor name-safe. This is becuase the kubectl command may have extra params that depends on the type of resource we are acting on. Doing this type-safe and name-safe would require to create different api for each resource. I found it too complex, and beside, if we ever migrate to lightkube, this may already be embedded there...